packages/x-date-pickers/src/internals/components/PickersArrowSwitcher/PickersArrowSwitcher.types.tsx

```diff
@@ -0,0 +1,89 @@
+import * as React from 'react';
+import { SlotComponentProps } from '@mui/base';
+import IconButton from '@mui/material/IconButton';
+import SvgIcon from '@mui/material/SvgIcon';
+
+export interface ExportedPickersArrowSwitcherProps {
+  /**
+   * Overrideable components.
+   * @default {}
+   */
+  components?: Partial<PickersArrowSwitcherSlotsComponent>;
+  /**
+   * The props used for each component slot.
+   * @default {}
+   */
+  componentsProps?: Partial<PickersArrowSwitcherSlotsComponentsProps>;
+  /**
+   * Left arrow icon aria-label text.
+   * @deprecated
+   */
+  leftArrowButtonText?: string;
+  /**
+   * Right arrow icon aria-label text.
+   * @deprecated
+   */
+  rightArrowButtonText?: string;
+}
+
+export interface PickersArrowSwitcherProps
+  extends ExportedPickersArrowSwitcherProps,
+    Omit<React.HTMLProps<HTMLDivElement>, 'ref' | 'as'> {
+  children?: React.ReactNode;
+  isLeftDisabled: boolean;
+  isLeftHidden?: boolean;
+  isRightDisabled: boolean;
+  isRightHidden?: boolean;
+  onLeftClick: () => void;
+  onRightClick: () => void;
+}
+
+export type PickersArrowSwitcherOwnerState = PickersArrowSwitcherProps;
+
+export interface PickersArrowSwitcherComponentsPropsOverrides {}
+
+export interface PickersArrowSwitcherSlotsComponent {
+  /**
+   * Button allowing to switch to the left view.
```

Comment:
```suggestion
   * Button allowing to switch to the previews view.
```

When left-to-right, is activated, is left the next view? using previous/next remove the confusion

Reply:
In this case we should probably rename the slots no ?

Reply:
If we still accept breaking changes, yes

Reply:
I think we will start the alpha quite soon, this PR can wait until then

Reply:
I tried to RTL mode and it's totally broken even on master
As far as I understand, we should not switch any icon / button but just the props associated to them.

I pushed a commit changing the behavior (and removing the deprecated locale text, there name were not correct and I don't think it was worth fixing them since we are merging this PR during the alpha phase)

Reply:
I'll second the opinion. If we are going with renames in the code as well as `localeText`, could we also rename the slots themselves? 🤔 
`PreviousArrowButton`, `NextArrowIcon` and so on.

Reply:
Also, couldn't we change the extension of this file to `ts`? 🤔

Reply:
From my testing of the RTL, I feel like the slots are named correctly

When I switch to RTL, I still want the arrow pointing left on the left and the arrow pointing right on the right.
But I want the arrow pointing left to go to the next view instead of the previous.

If we change the slots for "previous" and "next", then on RTL people should switch the icons to have something coherent displayed. Otherwise, they would have arrows pointing to one another like "> <" instead of "< >".

Unless my RTL test on Chrome was not representative of a real RTL setup

Reply:
![Screenshot 2022-09-02 at 12 00 21](https://user-images.githubusercontent.com/4941090/188104256-564f8372-633d-4877-a453-419a68c4eeff.png)
Toggling our docs theme setting does seem to act exactly like you said it shouldn't. 😢

Reply:
If you try on this PR, does it work correctly ?

I changed both the them setting and the RTL flag on Chrome to test, and it seems to work now

Reply:
I've tested on the PR docs version deployed on netlify. 🤔

Reply:
https://codesandbox.io/s/date-and-time-pickers-forked-wzyrmv?file=/package.json

Looks like your right
I followed [our guide](https://mui.com/material-ui/guides/right-to-left/#opting-out-of-rtl-transformation) to test on a real setup.
I must have failed my setup yesterday.

In that case, I'm also in favor of renaming the slots.

Reply:
As you saw I am not very confident on this RTL :laughing: 

But I'm not sure for the name of the slot
Let's take a concrete example: 

On LTR, on my screen I have:
- the 1st icon going to the previous view and looking like "<"
- the 2nd icon going to the next view and looking like ">"

If I switch to RTL, stylis changes the order of appearance on screen.
On the DOM I still have the same rendering order, but on my screen, it goes from right to left.

If we rename the `LeftArrowIcon` slot into `PreviousArrowIcon`.
Then the icon will be on the left on LTR and on the right on RTL => perfect.
But it will still be the "<" icon, to it will look like "> <"

So we would have to switch the icon default value depending on the direction.
AND users overriding it would have to do the same, which is not nice I think.


_____


I would rather keep the behavior existing before this PR.
Which is to switch in the code the order of rendering.

- on LTR, I render `{left} {right}` and it displays `{left} {right}` (or `< >` visually) 
- on RTL, I render `{right} {left}, but stylis invert on screen and displays `{left} {right}

Which is what we want, the left is still on the left with the RTL and if users want to switch the icon, they don't have to care about RTL to have something coherent visually in all circumstances.

But we switch the props applied when on RTL: we put the `next` props on the left and the `previous` props on the right.
That way, the icon on the left is going to next and the icon on the right to previous.

Reply:
Compared to my PR, it would be this diff:

```diff
--- a/packages/x-date-pickers/src/internals/components/PickersArrowSwitcher/PickersArrowSwitcher.tsx
+++ b/packages/x-date-pickers/src/internals/components/PickersArrowSwitcher/PickersArrowSwitcher.tsx
@@ -47,6 +47,7 @@ export const PickersArrowSwitcher = React.forwardRef(function PickersArrowSwitch
   ref: React.Ref<HTMLDivElement>,
 ) {
   const theme = useTheme();
+  const isRTL = theme.direction === 'rtl';
 
   const {
     children,
@@ -80,8 +81,7 @@ export const PickersArrowSwitcher = React.forwardRef(function PickersArrowSwitch
     label: previousLabel,
   };
 
-  const [leftProps, rightProps] =
-    theme.direction === 'rtl' ? [nextProps, previousProps] : [previousProps, nextProps];
+  const [leftProps, rightProps] = isRTL ? [nextProps, previousProps] : [previousProps, nextProps];
 
   const LeftArrowButton = components.LeftArrowButton ?? PickersArrowSwitcherButton;
   const leftArrowButtonProps = useSlotProps({
@@ -129,6 +129,18 @@ export const PickersArrowSwitcher = React.forwardRef(function PickersArrowSwitch
     ownerState: undefined,
   });
 
+  const leftPart = (
+    <LeftArrowButton {...leftArrowButtonProps}>
+      <LeftArrowIcon {...leftArrowIconProps} />
+    </LeftArrowButton>
+  );
+
+  const rightPart = (
+    <RightArrowButton {...rightArrowButtonProps}>
+      <RightArrowIcon {...rightArrowIconProps} />
+    </RightArrowButton>
+  );
+
   return (
     <PickersArrowSwitcherRoot
       ref={ref}
@@ -136,9 +148,7 @@ export const PickersArrowSwitcher = React.forwardRef(function PickersArrowSwitch
       ownerState={ownerState}
       {...other}
     >
-      <LeftArrowButton {...leftArrowButtonProps}>
-        <LeftArrowIcon {...leftArrowIconProps} />
-      </LeftArrowButton>
+      {isRTL ? rightPart : leftPart}
       {children ? (
         <Typography variant="subtitle1" component="span">
           {children}
@@ -146,9 +156,7 @@ export const PickersArrowSwitcher = React.forwardRef(function PickersArrowSwitch
       ) : (
         <PickersArrowSwitcherSpacer className={classes.spacer} ownerState={ownerState} />
       )}
-      <RightArrowButton {...rightArrowButtonProps}>
-        <RightArrowIcon {...rightArrowIconProps} />
-      </RightArrowButton>
+      {isRTL ? leftPart : rightPart}
     </PickersArrowSwitcherRoot>
   );
 });

```

Reply:
TLDR: On LTR the icon going to the next view looks like `>`, on RTL it looks like `<`.
So if we use previous/ next in the slot, the icon need to be direction-dependent which is not great for customizability.

Reply:
Another approach is to add a `noflip` like explained [here](https://mui.com/material-ui/guides/right-to-left/#emotion-amp-styled-components)

```diff
--- a/packages/x-date-pickers/src/internals/components/PickersArrowSwitcher/PickersArrowSwitcher.tsx
+++ b/packages/x-date-pickers/src/internals/components/PickersArrowSwitcher/PickersArrowSwitcher.tsx
@@ -16,9 +16,10 @@ const PickersArrowSwitcherRoot = styled('div', {
   overridesResolver: (props, styles) => styles.root,
 })<{
   ownerState: PickersArrowSwitcherProps;
-}>({
-  display: 'flex',
-});
+}>`
+  /* @noflip */
+  display: flex;
+`;
 
 const PickersArrowSwitcherSpacer = styled('div', {
   name: 'MuiPickersArrowSwitcher',
```

The core has only one `no-flip` in its codebase, maybe that's not something we want to rely on

Reply:
I've tested your [diff](https://github.com/mui/mui-x/pull/5890#discussion_r961499437) and to me it seems to work ideally 👍 
I've compared behaviour with https://demo.mobiscroll.com/javascript/calendar/rtl-right-to-left as well as Firefox native picker with Arabic language selected.
But after even more exploration I'm leaning 100% towards `Previous` and `Next` slot names instead of `Left` and `Right` to reduce confusion on what the user will actually be changing. 🤔

Reply:
> I'm leaning 100% towards Previous and Next slot names instead of Left and Right to reduce confusion on what the user will actually be changing. 

I'm confused
How do you solve the problems I'm describing above ?

> On LTR the icon going to the next view looks like >, on RTL it looks like <.
So if we use previous/ next in the slot, the icon need to be direction-dependent which is not great for customizability.

Reply:
@flaviendelangle I've done more looking into the `direction` conundrum and I'd be for not introducing too much "magic".
Basically, handle rendering different `arrows` based on `theme.direction` and if users will want to change icons and will need `rtl` support - they can also use `theme` hook to render whatever they need based on selected direction.
And let `css direction` handle rest - layout.
What do you [think about it](https://gist.github.com/LukasTy/3ad6334c4e69a1251a51b1423cf61d02)?
P.S. The diff also handles renaming `left/right` to `previous/next` to maintain prop "sanity" 😃

Reply:
I don't understand the point of going from left / right to prev / next if it make the override more verbous.

What is the magic introduced if we just call it left / right and assure it is rendered on the left and on the right with both LTR and RTL ?

Reply:
> assure it is rendered on the left and on the right with both LTR and RTL ?

How do you propose to switch the icons based on `direction`?
Rendering/layouting the buttons in the correct position is in charge of `direction: rtl`, we don't need to do anything regarding that, the issue is that arrows should be swapped based on `direction`.

Reply:
> How do you propose to switch the icons based on direction?

Why should we switch the icons ?
I feel like we are not talking about the same thing :laughing: 

My proposal is to keep everything like today:

- The slots are named `left` and `right`
- The left slot is always visible on the left side and the right slot is always visible on the right side (RTL or not)

The only modification I propose is maybe to use `@noflip` instead of a conditional rendering like we have on `master` to avoid playing with the rendering order.

And with this behavior, a developer wanting to change the icon, just change the icon, he does not have to be careful to do a ternary depending on the direction.

Reply:
> Why should we switch the icons ?

To maintain behaviour on which icon actually get's the provided props, because with the current approach and your suggested fix of flipping props it could get a bit confusing, where `rightArrowIcon` props would be applied either to the right (next) or left (previous) icon, whereas changing only the icon rendering would maintain behaviour of applying props to the same fundamental icon.
In the screenshots below I applied `title: 'Go to Previous'` to `leftArrowButton` and in RTL mode it would incorrectly apply this prop to icon button going to next month, instead of `previous`.
![Screenshot 2022-09-05 at 11 25 58](https://user-images.githubusercontent.com/4941090/188404331-9fbef72b-e4aa-4f46-972b-6748ed1867b8.png)
![Screenshot 2022-09-05 at 11 27 08](https://user-images.githubusercontent.com/4941090/188404335-5b36807a-c12f-43d2-812c-1ebf2e28ff39.png)

> My proposal is to keep everything like today:
The slots are named `left` and `right`

I don't have the strongest opinion on this point, slots can maintain the same names if we wish to reduce the amount of breaking changes, I'm simply in favour of them being renamed to align with all the other props and changes moving towards `previous/next` instead of `left/right` 🤔 

> The only modification I propose is maybe to use `@noflip` instead of a conditional rendering

Correct me if I'm wrong, but what which `rule` would we ask to not get `flipped` in `rtl`? We currently don't have any directional styles in pickers arrow switcher component. 🤔

Reply:
OK so basically, either we have a solution where developers can change the icon without caring about the direction, because the slot is bound to the position (left / right)

Or we have a solution where developers can change direction-related prop (like the title in your example) without caring about the direction, because the slot is bound to the action (go to previous / next)

____

I think I prefer to stick with the slot bound to the position.
Because it impacts one of the most basic customization: change the icon.

Whereas for the props, the title comes from the locale and should probably never be changed by prop.
And I don't see which prop you would like to customize that wants to know the current position of the icon (or button).
But maybe I'm missing some.

___

> where rightArrowIcon props would be applied either to the right (next) or left (previous) icon,

No, with my fix `rightArrowIcon` always go to the icon rendered on the right.
In LTR it's the next and on RTL it's the previous.
But it is always the one on the right.
Otherwise the naming would make no sense.
___

> I'm simply in favour of them being renamed to align with all the other props and changes moving towards previous/next instead of left/right 

But those are the props of `PickersArrowSwitcher`, not the props of `LeftArrowIcon`, `RightArrowIcon` and the 2 buttons.

Reply:
Not sure to get all the debate.

As far as I understand, it's about using right/left of prev/next. The problem is that with `rtl` the element order is modified and so the first element is put on the right instead of the left so you have to switch them to be consistent.

I took a look at the [`TablePagination`](https://github.com/mui/material-ui/blob/master/packages/mui-material/src/TablePagination/TablePaginationActions.js#L57-L66) in core which have the same problem. It sound like what they do is a distinction between the icon and the button.

For example, they have a `backIconButtonProps` (or `PreviousButton`) which render the following icon
```jsx
{theme.direction === 'rtl' ? <KeyboardArrowRight /> : <KeyboardArrowLeft />}
```

Having a `PreviouseIconButton`, `NextIconButton`, `ArrowLeftIcon`, and `ArrowRightIcon` could solve the problem because:
- buttons are always at the right place thanks to the browser `rtl` switch.
- `ArrowLeftIcon` defines an arrow pointing to the left instead of `LeftArrowIcon` which defines the arrow at the left of the component

Reply:
> ArrowLeftIcon defines an arrow pointing to the left instead of LeftArrowIcon which defines the arrow at the left of the component

So if we have others arrow icons elsewhere in the codebase of the pickers, would you share the slot ?
If not, the name is maybe too generic (and is already too generic today, I'm just thinking about that now).

The behavior you describe is OK for me.

Reply:
Sorry, I missed your answer

Sharing the slot is fine. For a given component, it sounds normal that the arrows to switch between views (or months) are similar

Reply:
Fixed
I'll check that it works in the full RTL setup

Reply:
Works fine https://codesandbox.io/s/date-and-time-pickers-forked-huxr57